src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala

```diff
@@ -39,47 +41,47 @@ case class IllegalLabelKeyException(labelKey: String)
 
 
-class LeonardoService(protected val dataprocConfig: DataprocConfig, protected val clusterResourcesConfig: ClusterResourcesConfig, protected val proxyConfig: ProxyConfig, gdDAO: DataprocDAO, dbRef: DbReference, val clusterMonitorSupervisor: ActorRef)(implicit val executionContext: ExecutionContext) extends LazyLogging {
+class LeonardoService(protected val dataprocConfig: DataprocConfig,
+                      protected val clusterResourcesConfig: ClusterResourcesConfig,
+                      protected val proxyConfig: ProxyConfig,
+                      protected val petServiceAccountConfig: PetServiceAccountConfig,
+                      protected val gdDAO: DataprocDAO,
+                      protected val googleIamDAO: GoogleIamDAO,
+                      protected val dbRef: DbReference,
+                      protected val clusterMonitorSupervisor: ActorRef,
+                      protected val samDAO: SamDAO)
+                     (implicit val executionContext: ExecutionContext) extends LazyLogging {
   private val bucketPathMaxLength = 1024
   private val includeDeletedKey = "includeDeleted"
 
   // Register this instance with the cluster monitor supervisor so our cluster monitor can potentially delete and recreate clusters
   clusterMonitorSupervisor ! RegisterLeoService(this)
 
-  def createCluster(googleProject: GoogleProject, clusterName: ClusterName, clusterRequest: ClusterRequest): Future[Cluster] = {
-    def create() = {
-      val clusterRequestWithDefaults = processClusterRequest(googleProject, clusterName, clusterRequest)
-      createGoogleCluster(googleProject, clusterName, clusterRequestWithDefaults) flatMap { case (cluster: Cluster, initBucket: GcsBucketName) =>
-        dbRef.inTransaction { dataAccess =>
-          dataAccess.clusterQuery.save(cluster, GcsPath(initBucket, GcsRelativePath("")))
-        }
-      }
+  def createCluster(userInfo: UserInfo, googleProject: GoogleProject, clusterName: ClusterName, clusterRequest: ClusterRequest): Future[Cluster] = {
+    samDAO.getPetServiceAccount(userInfo).flatMap { serviceAccount =>
+      createCluster(serviceAccount, googleProject, clusterName, clusterRequest)
     }
+  }
 
+  def createCluster(serviceAccount: WorkbenchUserServiceAccountEmail, googleProject: GoogleProject, clusterName: ClusterName, clusterRequest: ClusterRequest): Future[Cluster] = {
     // Check if the google project has a cluster with the same name. If not, we can create it
     dbRef.inTransaction { dataAccess =>
       dataAccess.clusterQuery.getActiveClusterByName(googleProject, clusterName)
     } flatMap {
       case Some(_) => throw ClusterAlreadyExistsException(googleProject, clusterName)
       case None =>
-        create andThen { case Success(cluster) =>
+        createGoogleCluster(serviceAccount, googleProject, clusterName, clusterRequest).flatMap { case (cluster, initBucket) =>
+          val augmentedCluster = addClusterDefaultLabels(cluster)
```

Comment:
I think if I try to create a cluster with a label with a forbidden key, it will create the cluster before checking the cluster request for forbidden keys.

Reply:
Ah hm that's true. Not sure why I moved it here - I think it could go before `createGoogleCluster` as before. Good catch!